File: src/17. Letter Combinations of a Phone Number/17.py
from typing import List


class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        conversion = {'2': 'abc', '3': 'def', '4': 'ghi', '5': 'jkl',
                      '6': 'mno', '7': 'pqrs', '8': 'tuv', '9': 'wxyz'}
        if len(digits) == 0:
            return []
        res = ['']
        for k in digits:
            for i in res:
                for j in conversion[k]:
                    res = res + [i+j]
                res.remove(i)
            # res=[i+j for i in res for j in conversion[k]]
        return res

    def letterCombinations1(self, digits: str) -> List[str]:
        phone = {'2': ['a', 'b', 'c'],
                 '3': ['d', 'e', 'f'],
                 '4': ['g', 'h', 'i'],
                 '5': ['j', 'k', 'l'],
                 '6': ['m', 'n', 'o'],
                 '7': ['p', 'q', 'r', 's'],
                 '8': ['t', 'u', 'v'],
                 '9': ['w', 'x', 'y', 'z']}

        def backtrack(combination, next_digits):
            if len(next_digits) == 0:
                output.append(combination)
            else:
                for letter in phone[next_digits[0]]:
                    backtrack(combination+letter, next_digits[1:])
        output = []
        if digits:
            backtrack("", digits)
        return output


if __name__ == "__main__":
    digits = "23"
    print(Solution().letterCombinations(digits))
    print(Solution().letterCombinations1(digits))

File: src/17. Letter Combinations of a Phone Number/17.go
var table []string = []string {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}

func letterCombinations(digits string) []string {
    if len(digits) == 0 {
        return []string{}
    }
    result := []string{""}
    for i := 0; i < len(digits); i++ {
        t := table[digits[i] - '0']
        temp := []string{}
        for j := 0; j < len(t); j++ {
            for z := 0; z < len(result); z++ {
                temp = append(temp, result[z] + string([]byte{t[j]}))
            }
        }
        result = temp
    }
    return result
}


var digitMap = map[string]string{"2":"abc", "3":"def", "4":"ghi", "5":"jkl", 
	"6":"mno", "7":"pqrs", "8":"tuv", "9":"wxyz"}
var combinations []string

func letterCombinations(digits string) []string {
	if len(digits)==0 {
        return nil
	}
	
    combinations = []string{}  // 重置全局变量
	getCombinations(digits, 0, "")
	return combinations
}

func getCombinations(digits string, index int, str string) {
	if index==len(digits) {
		combinations = append(combinations, str)
		return 
	}

	for _, s := range []rune(digitMap[digits[index:index+1]]) {
		getCombinations(digits, index+1, str+string(s))
	}
}

